### ISSUE TITLE
Amplify API | Subscription | Terminates with failed JSON parsing

### ISSUE THREAD

**Issue by meenakshi-rai** [aws-amplify/amplify-ios#1101]
I am attempting to get the `Subscription` to work with `Amplify API`, but every time as soon as it starts connecting, subscription soon terminates with following error
```
jsonParse(nil, Optional(Swift.DecodingError.dataCorrupted(Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "type", intValue: nil)], debugDescription: "Cannot initialize RealtimeConnectionProviderResponseType from invalid String value connection_error", underlyingError: nil))))
```

### Code Block
```
var subscription: GraphQLSubscriptionOperation<Chat>?

func subscribeToContent() {
        subscription = Amplify.API.subscribe(request: .subscription(of: Chat.self, type: .onCreate), valueListener: { subscriptionEvent in
            switch subscriptionEvent {
            case .connection(let subscriptionConnectionState):
                print("Subscription connect state is \(subscriptionConnectionState)") // Called 1
            case .data(let result):
                switch result {
                case .success(let createdMessage):
                    print("Successfully got response from subscription: \(createdMessage)")
                case .failure(let error):
                    print("Got failed result with \(error.errorDescription)")
                }
            }
        }, completionListener: { result in
            switch result {
            case .success:
                print("Subscription has been closed successfully")
            case .failure(let apiError):
                print("Subscription has terminated with \(apiError)") // Called 2
            }
        })
    }
```
#### Bug Description
1. Once this method gets called, it enters into `connection` case where it says `connecting`
2. After just a few seconds, it enters into `failure` case where it gives following error 
"jsonParse(nil, Optional(Swift.DecodingError.dataCorrupted(Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "type", intValue: nil)], debugDescription: "Cannot initialize RealtimeConnectionProviderResponseType from invalid String value connection_error", underlyingError: nil))))"

#### Expected Behavior
It should enter into `connection` case with value `connected`, and subscription should work

### Environment
Amplify Framework Version - 1.6.1
Dependency Manager - Cocoapods
Swift Version - 5
CLI Version - 4.45.0

### Device Information
Device - iPhone XS Max iOS 14.4 (device), iPhone SE 2nd Generation iOS 14.4 (simulator), iPhone Xr iOS 12.4 (simulator)

## Additional context
1. When we are using Authorization type as IAM `subscription FAILED`, but when we use API Key, `subscription WORKED`
2. Subscription works on Android and Web client using Amplify and same configuration

**Comment by lawmicha** [comment]
Hi @meenakshi-rai, could you provide more details around your set-up for us to reproduce?

1. What commands did you run in `amplify add api` and schema did you provide to set up your backend?
2. Are you able to successfully connect to your subscription API through AWS Appsync in the AWS Console?
3. What is in your Podfile, do you have AWSCognitoAuthPlugin?
4. Did you configure AWSCognitoAuthPlugin? API Key uses the key from the API plugin configuration, whereas IAM auth will use `Amplify.Auth` internally as part of the `Amplify.API` call.

**Comment by meenakshi-rai** [comment]
Hi @lawmicha , here are the details
## What commands did you run in amplify add api and schema did you provide to set up your backend?
### Commands used
```amplify init
amplify api add
amplify codegen models
```

### Schema used

```
type ChatConversation {
	id: ID!
	title: String
}

type ChatConnection {
	items: [AttendeeToAttendeeConversation]
	nextToken: String
}

type ChatConversationTest {
	id: ID!
	title: [String]
}

input CreateChatConversationInput {
	title: String
}

input CreateChatConversationInputTest {
	title: [String]
}

input DeleteChatConversationInput {
	id: ID!
}

type Mutation {
	createChatConversation(input: CreateChatConversationInput!): ChatConversation
	updateChatConversation(input: UpdateChatConversationInput!): ChatConversation
	deleteChatConversation(input: DeleteChatConversationInput!): ChatConversation
	createChatConversationTest(input: CreateChatConversationInputTest!): ChatConversationTest
}

type Query {
	getChatConversation(id: ID!): ChatConversation
	listChatConversations(filter: TableChatConversationFilterInput, limit: Int, nextToken: String): ChatConversationConnection
}

type Subscription {
	onCreateChatConversation: ChatConversation
		@aws_subscribe(mutations: ["createChatConversation"])
	onUpdateChatConversation(title: String): ChatConversation
		@aws_subscribe(mutations: ["updateChatConversation"])
	onDeleteChatConversation(id: ID): ChatConversation
		@aws_subscribe(mutations: ["deleteChatConversation"])
	onCreateChatConversationTest(title: String): ChateConversationTest
		@aws_subscribe(mutations: ["createChatConversationTest"])
}

input TableChatConversationFilterInput {
	id: TableIDFilterInput
	title: TableStringFilterInput
}

input TableBooleanFilterInput {
	ne: Boolean
	eq: Boolean
}

input TableFloatFilterInput {
	ne: Float
	eq: Float
	le: Float
	lt: Float
	ge: Float
	gt: Float
	contains: Float
	notContains: Float
	between: [Float]
}

input TableIDFilterInput {
	ne: ID
	eq: ID
	le: ID
	lt: ID
	ge: ID
	gt: ID
	contains: ID
	notContains: ID
	between: [ID]
	beginsWith: ID
}

input TableIntFilterInput {
	ne: Int
	eq: Int
	le: Int
	lt: Int
	ge: Int
	gt: Int
	contains: Int
	notContains: Int
	between: [Int]
}

input TableStringFilterInput {
	ne: String
	eq: String
	le: String
	lt: String
	ge: String
	gt: String
	contains: String
	notContains: String
	between: [String]
	beginsWith: String
}

input UpdateChatConversationInput {
	id: ID!
	title: String
}

```

## Are you able to successfully connect to your subscription API through AWS Appsync in the AWS Console?
Yes

## What is in your Podfile, do you have AWSCognitoAuthPlugin?
```
# Pods for Amplify
  pod 'Amplify', '~> 1.6.1'
  pod 'AmplifyPlugins/AWSCognitoAuthPlugin', '~> 1.6.1'
  pod 'AmplifyPlugins/AWSAPIPlugin', '~> 1.6.1'
```

## Did you configure AWSCognitoAuthPlugin? API Key uses the key from the API plugin configuration, whereas IAM auth will use Amplify.Auth internally as part of the Amplify.API call.
- Yes AWSCognitoAuthPlugin is used and configured in code. 
- Here is the structure of amplfyconfiguration.json

```
{
  "UserAgent": "aws-amplify-cli/2.0",
  "Version": "1.0",
  "api": {
    "plugins": {
      "awsAPIPlugin": {
        "CNM": {
          "endpointType": "GraphQL",
          "endpoint": "https://xxxxxxxxxxxxxxxxxxxxxxxxxx.appsync-api.us-east-1.amazonaws.com/graphql",
          "region": "us-east-1",
          "authorizationType": "AWS_IAM"
        }
      }
    }
  },
  "auth": {
    "plugins": {
      "awsCognitoAuthPlugin": {
        "UserAgent": "aws-amplify/cli",
        "Version": "0.1.0",
        "IdentityManager": {
          "Default": {}
        },
        "CredentialsProvider": {
          "CognitoIdentity": {
            "Default": {
              "PoolId": "us-east-1:1234a123-1234-1234-abcd-123456789101",
              "Region": "us-east-1"
            }
          }
        },
        "Auth": {
          "Default": {
            "authenticationFlowType": "USER_SRP_AUTH"
          }
        },
        "AppSync": {
          "Default": {
            "ApiUrl": "https://xxxxxxxxxxxxxxxxxxxxxxxxxx.appsync-api.us-east-1.amazonaws.com/graphql",
            "Region": "us-east-1",
            "AuthMode": "AWS_IAM"
          }
        }
      }
    }
  }
}
```

**Comment by meenakshi-rai** [comment]
Hi @lawmicha, 
Based on our test this does not look like an issue on AppSync side. 
We are using `Dynatrace monitoring`, and for subscription request, the signed header list contains `x-dynatrace` header, but the same header is not passed as part of authorization header, which makes the signature validation to fail on AppSync side. If we disable `Dynatrace` then it starts working.

#### Sample Header in a subscription request using IAM, returned in `IAMAuthInterceptor.swift`

```
{
authorization = "AWS4-HMAC-SHA256 Credential=XXXXX/12345/us-east-1/appsync/aws4_request, SignedHeaders=accept;content-encoding;content-type;host;x-amz-date;x-amz-security-token;x-dynatrace, Signature=XXXXXXXXXXXXXXXXXXXXXX"
  securityToken = "xxx"
  date = "20210316T055843Z"
  accept = "application/json, text/javascript"
  contentEncoding = "amz-1.0"
  contentType = "application/json; charset=UTF-8"
}
```

class named `IAMAuthenticationHeader`, which is a decodable model for hard-coded header's, but the `SignedHeaders` use all the headers being injected, so that is where the variation is.

**Comment by lawmicha** [comment]
Hi @meenakshi-rai, thanks for the details. I can see that the sigv4 signer is generating `Authorization` header with `SignedHeaders=headerKey1,headerkey2,headerKey3` string from all header fields from the original request. ([AWSSignature.m](https://github.com/aws-amplify/aws-sdk-ios/blob/main/AWSCore/Authentication/AWSSignature.m#L362))

The `Authorization` header is then apart of this header object, as you mentioned represented by the class [IAMAuthenticationHeader](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift#L111)

IAMAuthenticationHeader has specific coding keys which will generate a JSON object as
```
{
    "accept": "application/json, text/javascript",
    "content-encoding": "amz-1.0",
    "content-type": "application/json; charset=UTF-8",
    "host": "xxx.appsync-api.region.amazonaws.com",
    "x-amz-date": "20210316T055843Z",
    "X-Amz-Security-Token": "xxx",
    "Authorization": "AWS4-HMAC-SHA256 Credential=xxxxxx/aws4_request, SignedHeaders=accept;content-encoding;content-type;host;x-amz-date;x-amz-security-token, Signature=xxxxxx"
}
```
This header object gets base64 encoded to a string, and added to the query parameter "header" in the final subscription connection request. 
```
wss://xxxx.appsync-realtime-api.region.amazonaws.com/graphql?header=[base64encoded header object]
```

It appears in code that the SigV4 signer is using the `x-dynatrace` header, by 
1. creating a canonical request from all header fields (probably including `x-dynatrace`)
2. creating a string to sign from the canonical request
3. creating a signature using sha256 and key
4. generating the "Signature=xxx" value, added to the  `Authorization` header. 
5. generating the "SignedHeaders=xxx" value, added to the `Authorization` header. 

I wonder if `x-dynatrace` should be used to generate the SigV4 signature, and thus added to the header object, or if it can simply be apart of the final request as a header field, that is not signed.

Since this is working on Android and Web, do you know if the `x-dynatrace` is being added to the `SignedHeader` for the other platforms? How is `x-dynatrace` header added to original `GraphQLRequest` in the first place?

**Comment by meenakshi-rai** [comment]
Hi @lawmicha,

Dynatrace header field `x-dynatrace` is auto-injected field, and is injected in each URLRequest on it’s own.
 
Here is how the mutation create (GraphQLReuqest) request looks like, with Dynatrace header attached –
 
▿ url : Optional<URL>
    ▿ some : https://xxxxx.appsync-api.us-east-1.amazonaws.com/graphql
      - _url : https://xxxxx.appsync-api.us-east-1.amazonaws.com/graphql
  - cachePolicy : 0
  - timeoutInterval : 60.0
  - mainDocumentURL : nil
  - networkServiceType : __C.NSURLRequestNetworkServiceType
  - allowsCellularAccess : true
  ▿ httpMethod : Optional<String>
    - some : "POST"
  ▿ allHTTPHeaderFields : Optional<Dictionary<String, String>>
    ▿ some : 8 elements
      ▿ 0 : 2 elements
        - key : "X-Amz-Security-Token"
        - value : "XXXXXXXXXXXX"
      ▿ 1 : 2 elements
        - key : "User-Agent"
        - value : "amplify-iOS/1.6.1 iOS/12.4 en_US"
      ▿ 2 : 2 elements
        - key : "Authorization"
        - value : "AWS4-HMAC-SHA256 Credential=XXXXX/12344/us-east-1/appsync/aws4_request, SignedHeaders=content-type;host;user-agent;x-amz-date;x-amz-security-token;x-dynatrace, Signature=qw35252532gf2t3y534y35363t"
      ▿ 3 : 2 elements
        - key : "accessToken"
        - value : "abcdefghi-1234-1234-1234-abcdefghijkl"
      ▿ 4 : 2 elements
        - key : "x-dynatrace"
        - value : "XXXXXXXXXXXXXXXXXXXX"
      ▿ 5 : 2 elements
        - key : "Host"
        - value : "xxxxxxx.appsync-api.us-east-1.amazonaws.com"
      ▿ 6 : 2 elements
        - key : "Content-Type"
        - value : "application/json"
      ▿ 7 : 2 elements
        - key : "X-Amz-Date"
        - value : "20210316T093910Z"
  ▿ httpBody : Optional<Data>
    ▿ some : 218 bytes
      - count : 218
      ▿ pointer : 0x00007ff96e156000
        - pointerValue : 12234235346346346
  - httpBodyStream : nil
  - httpShouldHandleCookies : true
  - httpShouldUsePipelining : false

NOTE - On Android, “x-dynatrace” is NOT added in SignedHeaders list and skipped for signing as well making validation a success at Aws end.

**Comment by meenakshi-rai** [comment]
Hi @lawmicha, 
In iOS we use cocoapods library and on Android, gradle build library AND, enable auto-instrumentation feature for it. We do not inject these fields, manually. Below is their documented process.
 
**How Dynatrace iOS Works**
Dynatrace OneAgent for iOS automatically times web requests that are made using NSURLRequest, NSURLConnection, NSURLSession, NSURLProtocol, NSString, or WKWebView.
Same details can be found under this document (Measure web requests)
**NOTE** - Since the SDK is not open-sourced it is difficult to know their absolute technique, but most likely being done using NSURLProtocol, while making request
https://www.dynatrace.com/support/help/technology-support/operating-systems/ios/customization/oneagent-sdk-for-ios/
 
**How Dynatrace Android Works**
The following HTTP frameworks are supported:

1. HttpURLConnection
2. OkHttp: Only version 3 and 4
3. Apache HttpClient: Only the Android-internal HttpClient version 
If your web request library is based on one of these supported frameworks, the internal classes of the library are automatically instrumented.

**NOTE** - Since the SDK is not open-sourced it is difficult to know their absolute technique, but most likely being done using network interceptors
https://www.dynatrace.com/support/help/shortlink/dynatrace-android-gradle-plugin-monitoring#web-request-monitoring-sensors
 

> I'm curious, in the example of the create mutation you have above, it contains accessToken, which process/how is this being added?

We use an extension provided by `AWSAPIPlugin`, which allows to add an interceptor to pass any header in internal URLRequest being made. We use it to pass a token, which connects AppSync storage to our own database to identify the user. 
 Code mentioned below:
```
let apiPlugin = AWSAPIPlugin()
try Amplify.add(plugin: apiPlugin)
try Amplify.configure()            
try apiPlugin.add(interceptor: AmplifyRequestInterceptor(), for: AmplifyConstants.API.name)
 
/// Custom Struct used to intercept URLRequest within GraphQL API call using Amplify API library
struct AmplifyRequestInterceptor: URLRequestInterceptor {
    
    /// Intercepts any internal URLRequest made by Amplify Library for GraphQL and mutates request as per definition
    /// - Parameter request: URLRequest intercepted while a grapqhQL API call is made
    /// - Throws: APIError thrown if URLRequest is not valid
    /// - Returns: Modified URLRequest to be sent to server
    func intercept(_ request: URLRequest) throws -> URLRequest {
 
        guard let mutableRequest = (request as NSURLRequest).mutableCopy() as? NSMutableURLRequest else {
            throw APIError.unknown("Could not get mutable request", "")
        }
        // Place where access token get’s added. It let’s us connect with our own database
        let accessToken = AuthorizationManager.sharedInstance.getUserToken()?.token ?? ""
        mutableRequest.setValue(accessToken, forHTTPHeaderField: AmplifyConstants.API.headerAccessToken)
        return mutableRequest as URLRequest
    }
}
```

**Comment by lawmicha** [comment]
Hi @meenakshi-rai, thanks for the links. It sounds like the auto-instrumentation feature has a difference across platforms. If this is working on Android then the dynatrace header must be added after sigv4 signing occurs. On iOS, for a subscription request, the `x-dynatrace` header is added to the request before it is signed, then the signature is stored in the Authorization header, then the entire set of headers (excluding `x-dynatrace`, including the Authorization header) is encoded as the payload in the "header" query string. For a query/mutation request, `x-dynatrace` is added to the request headers, then the request is signed which generates the Authorization header. 

What I can think of so far..
1. Could you disable dynatrace's auto-instrumentation feature for just subscriptions? Is it critical to monitor subscription connections using dynatrace?
2. If not, could you disable dynatrace auto-instrumentation feature all together and add the dynatrace header similar to how you add the access token? This is to avoid the SigV4 signing against the `x-dynatrace` header and have it added to the request after signing occurs. Since the interceptor mechanism only applies to the mutation and query requests, we would have to build a similar subscription connection request interceptor to support this.
3. Another option that would require no changes from you, because a subsccription request is created using an `IAMAuthenticationHeader` object and has access to the headers from the original request ([code](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift#L97-L98)), it could check `mutableRequest.allHTTPHeaderFields` to gather additional header values to store in the `authHeader` (payload for the "header" query string) by creating a more flexible `IAMAuthenticationHeader` like object. This would require some testing to ensure that it doesn't break anything, actually solves your use case, and the AppSync service allows for extra headers in the "header" query string. This brings up another question regarding the subscription request on Android, is the `x-dynatrace` header being added to the request's headers or being encoded in the "header" query string?

**Comment by Max-Rai** [comment]
Hi @lawmicha, 

1. I would suggest to go with step 3, as we did not found options in Dynatrace for iOS, to stop auto-instrumentation for specific requests
2. Adding Dynatrace header similar to accesstoken also I am not sure about as we have no clue, how Dynatrace generates the header value
3. I have not seen in Android, any same header passed of Dynatrace in subscription, so can’t comment how Dynatrace or IF Dynatrace is instrumenting Android AWS API call’s

**Comment by meenakshi-rai** [comment]
Hi @lawmicha,

1. I would suggest to go with step 3, as we did not found options in Dynatrace for iOS, to stop auto-instrumentation for specific requests
2. Adding Dynatrace header similar to accesstoken also I am not sure about as we have no clue, how Dynatrace generates the header value
3. I have not seen in Android, any same header passed of Dynatrace in subscription, so can’t comment how Dynatrace or IF Dynatrace is instrumenting Android AWS API call’s

**Comment by lawmicha** [comment]
Hi @meenakshi-rai, thanks for following up. I have a draft PR that implements option 3: https://github.com/aws-amplify/amplify-ios/pull/1139/files I do need to do some additional testing to ensure this there isn't any regression and to get the team's review on the proposed changes.

Since you have dynatrace SDK set up on your side, could you verify that this change will solve your problem? 

Install the pods at the branch
```
pod 'Amplify', :git => 'https://github.com/aws-amplify/amplify-ios.git', :branch => 'fix/subscribe-iam-header'
  pod 'AmplifyPlugins/AWSAPIPlugin', :git => 'https://github.com/aws-amplify/amplify-ios.git', :branch => 'fix/subscribe-iam-header'
  pod 'AmplifyPlugins/AWSCognitoAuthPlugin', :git => 'https://github.com/aws-amplify/amplify-ios.git', :branch => 'fix/subscribe-iam-header'
```

And perform your use case, ie. `Amplify.API.subscribe` and any other related calls.

If you are able to test the change out, could you also break point when the NSMutableRequest is created [here](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift#L81)?

Once you step over that, you should be able to print out the values of the `mutableRequest` with the console with `po mutableRequest.allHeaderValues`. Is the `x-dynatrace` header added there after this line?

**Comment by meenakshi-rai** [comment]
Hi, @lawmicha 
1. I enabled Dynatrace and confirmed that with this check, subscription was able to enter into `connected` state
2. I added breakpoint in `IAMAuthInterceptor`, after `mutableRequest` was created [here](https://github.com/aws-amplify/amplify-ios/blob/fix/subscribe-iam-header/AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift#L81), and this is what I received back using `print mutableRequest.allHTTPHeaderFields`, it does contain `x-dynatrace` header
 
([String : String]?) $R0 = 5 key/value pairs {
  [0] = (key = "Content-Type", value = "application/json; charset=UTF-8")
  [1] = (key = "x-dynatrace", value = "MT_3_1_123456789_1_abcddesd-12sd-cd56-ka78-jhdf989ajbh_0_12_878")
  [2] = (key = "Content-Encoding", value = "amz-1.0")
  [3] = (key = "x-amz-date", value = "20210405T103752Z")
  [4] = (key = "Accept", value = "application/json, text/javascript")
}

**Comment by lawmicha** [comment]
Hi @meenakshi-rai, thanks for your patience, https://github.com/aws-amplify/amplify-ios/pull/1139 has been merged to main, and will be released this week. Will keep you posted

**Comment by lawmicha** [comment]
The changes have been released in 1.8.1 https://github.com/aws-amplify/amplify-ios/releases/tag/v1.8.1 please let me know if you have any questions
